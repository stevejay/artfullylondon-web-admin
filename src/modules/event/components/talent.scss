@import 'theme';

.container {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.img {
  width: 60px;
  height: 60px;
  display: block;
}

.img,
.placeholder {
  border-radius: 500rem;
}

.details {
  display: flex;
  flex-direction: column;
  margin-left: 1rem;

  h4,
  .roles,
  .characters {
    margin: 0;
    padding: 0;
  }

  h4 {
    @include regularFont;
    font-size: 1.6rem;
  }

  .roles {
    @include extraBoldFont;
    text-transform: uppercase;
    font-size: 1.2rem;
  }

  .characters {
    font-style: italic;
  }
}
