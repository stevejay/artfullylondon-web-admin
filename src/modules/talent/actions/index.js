export const types = {
  CREATE_TALENT_FOR_ENTITY: 'talent/CREATE_TALENT_FOR_ENTITY'
}

export const createTalentForEntity = (values, parentFormName) => ({
  type: types.CREATE_TALENT_FOR_ENTITY,
  payload: {
    values,
    parentFormName
  },
  meta: { thunk: true }
})
