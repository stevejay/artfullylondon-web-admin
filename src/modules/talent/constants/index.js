import talentType from '_src/entities/types/talent-type'

export const EDIT_TALENT_FORM_NAME = 'EditTalent'

export const TALENT_TYPES_DROPDOWN_OPTIONS = [
  {
    value: talentType.INDIVIDUAL,
    label: talentType.INDIVIDUAL
  },
  {
    value: talentType.GROUP,
    label: talentType.GROUP
  }
]

export const TALENT_CONSTRAINT = {
  firstNames: {
    length: { maximum: 200 }
  },
  lastName: {
    presence: { disallowEmpty: true },
    length: { maximum: 200 }
  },
  commonRole: {
    presence: { disallowEmpty: true },
    length: { maximum: 200 }
  },
  images: {
    length: {
      maximum: 10,
      tooLong: 'has too many elements'
    }
  },
  weSay: {
    length: { maximum: 300 }
  }
}

export const TALENT_NORMALISER = {
  firstNames: {
    trim: true
  },
  lastName: {
    trim: true
  },
  commonRole: {
    trim: true
  },
  weSay: {
    trim: true
  }
}
