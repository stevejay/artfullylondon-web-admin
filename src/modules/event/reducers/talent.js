import { handleActions } from 'redux-actions'

import { types } from '../actions'

const initialState = { talentId: null }

export const reducer = handleActions(
  {
    [types.UPDATE_SELECTED_TALENT]: (state, action) => ({
      ...state,
      talentId: action.payload.talentId
    })
  },
  initialState
)

export const selectors = {
  selectedTalentId: state => state.talentId
}
