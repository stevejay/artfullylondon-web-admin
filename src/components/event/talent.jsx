import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import { ENTITY_TYPE_TALENT } from '_src/constants/entity'
import ImagePlaceholder from '_src/components/image-placeholder'
import * as image from '_src/lib/image'
import { SummaryTalent } from '_src/entities/talent'
import './talent.m.scss'

class EventTalentComponent extends React.Component {
  shouldComponentUpdate (nextProps) {
    return nextProps.talent !== this.props.talent
  }
  render () {
    const { talent, ...rest } = this.props

    return (
      <div {...rest} styleName='container'>
        {talent.hasImage &&
          <img
            styleName='img'
            src={image.createEventTalentImageUrl(talent.image)}
          />}
        {!talent.hasImage &&
          <ImagePlaceholder
            styleName='placeholder'
            size='very-small'
            type={ENTITY_TYPE_TALENT}
          />}
        <div styleName='details'>
          <p styleName='roles'>{talent.createRolesString()}</p>
          {talent.hasCharacters() &&
            <p styleName='characters'>
              {talent.createCharactersString()}
            </p>}
          <h4>
            <Link to={talent.url}>{talent.name}</Link>
          </h4>
        </div>
      </div>
    )
  }
}

EventTalentComponent.propTypes = {
  talent: PropTypes.instanceOf(SummaryTalent).isRequired
}

export default EventTalentComponent
