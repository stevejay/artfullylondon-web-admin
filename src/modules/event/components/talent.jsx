import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

import { ImagePlaceholder } from '_src/modules/image'
import * as entityConstants from '_src/constants/entity'
import * as image from '_src/lib/image'
import { SummaryTalent } from '_src/entities/talent'
import './talent.scss'

class EventTalentComponent extends React.PureComponent {
  render () {
    const { talent, className, ...rest } = this.props

    return (
      <div {...rest} className={className} styleName='container'>
        {talent.hasImage &&
          <img
            styleName='img'
            src={image.createEventTalentImageUrl(talent.image)}
          />}
        {!talent.hasImage &&
          <ImagePlaceholder
            styleName='placeholder'
            size='very-small'
            type={entityConstants.ENTITY_TYPE_TALENT}
          />}
        <div styleName='details'>
          <p styleName='roles'>{talent.createRolesString()}</p>
          {talent.hasCharacters() &&
            <p styleName='characters'>
              {talent.createCharactersString()}
            </p>}
          <h4>
            <Link to={talent.url}>{talent.name}</Link>
          </h4>
        </div>
      </div>
    )
  }
}

EventTalentComponent.propTypes = {
  className: PropTypes.string,
  talent: PropTypes.instanceOf(SummaryTalent).isRequired
}

export default EventTalentComponent
