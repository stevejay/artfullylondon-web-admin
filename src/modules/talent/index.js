import TalentEditOrCreate from './components/talent-edit-or-create'
import TalentDetail from './components/talent-detail'
import TalentsField from './components/talents-field'
import CreateBasicTalentForm from './forms/create-basic-talent'
import { getBasicTalentInitialValues } from './lib/mapper'
import sagas from './sagas'
import * as actions from './actions'

export {
  TalentEditOrCreate,
  TalentDetail,
  TalentsField,
  CreateBasicTalentForm,
  getBasicTalentInitialValues,
  sagas,
  actions
}
