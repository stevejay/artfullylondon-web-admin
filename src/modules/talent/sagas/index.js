import { call, put, takeLatest } from 'redux-saga/effects'
import { startSubmit, stopSubmit, reset, arrayPush } from 'redux-form'
import log from 'loglevel'

import * as sagaLib from '_src/shared/lib/saga'
import * as validationLib from '_src/shared/lib/validation'
import * as talentActions from '../actions'
import * as talentConstants from '../constants'
import * as talentMapper from '../lib/mapper'
import entityType from '_src/domain/types/entity-type'
import normalise from '_src/shared/lib/normalise'
import { eventService } from '_src/modules/api'
import { actions as notificationActions } from '_src/modules/notification'

export function * createTalentForEntity ({
  payload: { values: payloadValues, parentFormName },
  meta
}) {
  try {
    yield put(startSubmit(talentConstants.EDIT_TALENT_FORM_NAME))

    const values = yield call(
      normalise,
      payloadValues,
      talentConstants.TALENT_NORMALISER
    )

    yield call(
      validationLib.validate,
      values,
      talentConstants.TALENT_CONSTRAINT
    )

    const entity = yield call(
      eventService.saveEntity,
      entityType.TALENT,
      values,
      talentMapper.mapSubmittedValues,
      false
    )

    const newTalent = {
      ...entity,
      key: entity.id,
      roles: entity.commonRole,
      characters: ''
    }

    yield put(arrayPush(parentFormName, 'talents', newTalent))
    yield put(stopSubmit(talentConstants.EDIT_TALENT_FORM_NAME))
    yield put(reset(talentConstants.EDIT_TALENT_FORM_NAME))
    yield put(sagaLib.returnAsPromise(null, meta))
  } catch (err) {
    yield call(log.error, err)

    yield put(
      notificationActions.addErrorNotification('Save Failed', err.message)
    )

    yield call(
      sagaLib.submitErrorHandler,
      err,
      talentConstants.EDIT_TALENT_FORM_NAME
    )
  }
}

export default [
  takeLatest(
    talentActions.types.CREATE_TALENT_FOR_ENTITY,
    createTalentForEntity
  )
]
