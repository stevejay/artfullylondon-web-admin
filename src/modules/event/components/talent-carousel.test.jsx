import React from 'react'
import _ from 'lodash'

import EventTalentCarousel from './talent-carousel'
import { EventSummaryTalent } from '_src/entities/talent'
import IconButton from '_src/components/button/icon'

it('should render correctly', () => {
  const wrapper = shallow(
    <EventTalentCarousel
      talents={[
        new EventSummaryTalent({ id: 'talent-1' }),
        new EventSummaryTalent({ id: 'talent-2' })
      ]}
      selectedTalentId='talent-1'
      onTalentSelected={_.noop}
    />
  )

  expect(wrapper).toMatchSnapshot()
})

it('should render correctly when the selected talent id is null', () => {
  const wrapper = shallow(
    <EventTalentCarousel
      talents={[
        new EventSummaryTalent({ id: 'talent-1' }),
        new EventSummaryTalent({ id: 'talent-2' })
      ]}
      selectedTalentId={null}
      onTalentSelected={_.noop}
    />
  )

  expect(wrapper).toMatchSnapshot()
})

it('should render correctly when there is only one talent', () => {
  const wrapper = shallow(
    <EventTalentCarousel
      talents={[new EventSummaryTalent({ id: 'talent-1' })]}
      selectedTalentId='talent-1'
      onTalentSelected={_.noop}
    />
  )

  expect(wrapper).toMatchSnapshot()
})

it('should render correctly when there are no talents', () => {
  const wrapper = shallow(
    <EventTalentCarousel talents={[]} onTalentSelected={_.noop} />
  )

  expect(wrapper).toMatchSnapshot()
})

it('should handle a left arrow click when there is an image to the left', () => {
  const handleTalentSelected = jest.fn()

  const wrapper = shallow(
    <EventTalentCarousel
      talents={[
        new EventSummaryTalent({ id: 'talent-1' }),
        new EventSummaryTalent({ id: 'talent-2' })
      ]}
      selectedTalentId='talent-2'
      onTalentSelected={handleTalentSelected}
    />
  )

  wrapper.find(IconButton).at(1).simulate('click')

  expect(handleTalentSelected).toHaveBeenCalledWith({ talentId: 'talent-1' })
})

it('should handle a left arrow click when it should wrap around', () => {
  const handleTalentSelected = jest.fn()

  const wrapper = shallow(
    <EventTalentCarousel
      talents={[
        new EventSummaryTalent({ id: 'talent-1' }),
        new EventSummaryTalent({ id: 'talent-2' })
      ]}
      selectedTalentId='talent-1'
      onTalentSelected={handleTalentSelected}
    />
  )

  wrapper.find(IconButton).at(1).simulate('click')

  expect(handleTalentSelected).toHaveBeenCalledWith({ talentId: 'talent-2' })
})

it('should handle a right arrow click when there is an image to the right', () => {
  const handleTalentSelected = jest.fn()

  const wrapper = shallow(
    <EventTalentCarousel
      talents={[
        new EventSummaryTalent({ id: 'talent-1' }),
        new EventSummaryTalent({ id: 'talent-2' })
      ]}
      selectedTalentId='talent-1'
      onTalentSelected={handleTalentSelected}
    />
  )

  wrapper.find(IconButton).at(0).simulate('click')

  expect(handleTalentSelected).toHaveBeenCalledWith({ talentId: 'talent-2' })
})

it('should handle a right arrow click when it should wrap around', () => {
  const handleTalentSelected = jest.fn()

  const wrapper = shallow(
    <EventTalentCarousel
      talents={[
        new EventSummaryTalent({ id: 'talent-1' }),
        new EventSummaryTalent({ id: 'talent-2' })
      ]}
      selectedTalentId='talent-2'
      onTalentSelected={handleTalentSelected}
    />
  )

  wrapper.find(IconButton).at(0).simulate('click')

  expect(handleTalentSelected).toHaveBeenCalledWith({ talentId: 'talent-1' })
})
