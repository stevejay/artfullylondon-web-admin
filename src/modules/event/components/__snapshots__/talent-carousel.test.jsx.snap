// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="src-modules-event-components-talent-carousel__container"
>
  <EventTalentComponent
    className="src-modules-event-components-talent-carousel__event-talent"
    talent={
      SummaryTalent {
        "entity": Object {
          "id": "talent-1",
        },
      }
    }
  />
  <div
    className="src-modules-event-components-talent-carousel__buttons-container"
  >
    <IconButton
      aria-label="Next talent"
      icon={[Function]}
      onClick={[Function]}
      type="default"
    />
    <IconButton
      aria-label="Previous talent"
      icon={[Function]}
      onClick={[Function]}
      type="default"
    />
  </div>
</div>
`;

exports[`should render correctly when the selected talent id is null 1`] = `
<div
  className="src-modules-event-components-talent-carousel__container"
>
  <EventTalentComponent
    className="src-modules-event-components-talent-carousel__event-talent"
    talent={
      SummaryTalent {
        "entity": Object {
          "id": "talent-1",
        },
      }
    }
  />
  <div
    className="src-modules-event-components-talent-carousel__buttons-container"
  >
    <IconButton
      aria-label="Next talent"
      icon={[Function]}
      onClick={[Function]}
      type="default"
    />
    <IconButton
      aria-label="Previous talent"
      icon={[Function]}
      onClick={[Function]}
      type="default"
    />
  </div>
</div>
`;

exports[`should render correctly when there are no talents 1`] = `""`;

exports[`should render correctly when there is only one talent 1`] = `
<EventTalentComponent
  talent={
    SummaryTalent {
      "entity": Object {
        "id": "talent-1",
      },
    }
  }
/>
`;
